enhancements/installer/signal-cluster-deletion.md

```diff
@@ -0,0 +1,121 @@
+---
+title: signal-cluster-deletion
+authors:
+  - "@abutcher"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-03-30
+last-updated: 2020-03-30
+status: provisional
+---
+
+# Signal Cluster Deletion
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Operators can create external resources in clouds and other services that have no opportunity to be cleaned up when a cluster is uninstalled, causing extra work for cluster maintainers to track down and remove these resources. Signalling cluster deletion would allow operators to clean up external resources during uninstallation and would ensure that total cleanup is an automatic process.
+
+## Motivation
+
+### Goals
+
+Signal that cluster deletion has begun through deletion of a new CRD `ClusterAlive` and wait for that object to be deleted during uninstall before continuing with removing cluster resources such as instances, storage and networking.
+
+### Non-Goals
+
+Attaching finalizers to operator resources based on the `ClusterAlive` object which would facilitate removal of operator resources.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints
+
+#### API
+
+Introduce a new CRD `ClusterAlive` from which operator resources could attach a finalizer initiating cascading delete when the `ClusterAlive` resource is deleted.
+
+```yaml
+apiVersion: v1
+kind: ClusterAlive
+metadata:
+  name: cluster
+  namespace: openshift-config
+```
+
+#### Install
+
+Create a `ClusterAlive` object during installation. (Is a manifest the right place to create for new clusters ?)
+
+#### Upgrade
+
+Create `ClusterAlive` object for existing clusters. (Is this the right place to create for existing clusters ?)
+
+#### Uninstall
+
+Delete a cluster's `ClusterAlive` resource during uninstallation and wait X minutes before continuing cluster tear down.
```

Comment:
Should we add specifics here? 

Proposal: modify openshift-install destory cluster to do this. Attempt to delete ClusterAlive by default, provide a flag to opt-out. If the cluster is unreachable we error, if the timeout is hit we error, and the user must run with the opt-out flag. 

--skip-clusteralive-delete
--clusteralive-delete-timeout 30
--clusteralive-delete-failure-policy continue|abort

This would help Hive as I believe we're going to want to always try this, but continue if the timeout is hit or we can't reach the cluster.

Reply:
I expect some pushback to "on by default" uninstall behavior but I like these recommendations.

Reply:
>Proposal: modify openshift-install destory cluster to do this. Attempt to delete ClusterAlive by default, provide a flag to opt-out. If the cluster is unreachable we error, if the timeout is hit we error, and the user must run with the opt-out flag.
>
>--skip-clusteralive-delete
--clusteralive-delete-timeout 30
--clusteralive-delete-failure-policy continue|abort

currently users expect the destroy to work, even in cases where the cluster api is not running. with your proposal the usres will now error if the api is unavailable, or this new delete fails. to me that sounds like backward-incompatible change.

Reply:
> --skip-clusteralive-delete
--clusteralive-delete-timeout 30
--clusteralive-delete-failure-policy continue|abort

secondly, the openshift destroy command is supposed to be idempontent, so instead of giving users timeout option, i would ask the users be expected to call destroy multiple times.

third, continue/abort is not something we should support, either the users can skip deletion of clusteralive or keep retrying until success. this continue/abort is foot gun.

Reply:
What would you expect CI to do on teardown if ClusterAlive gets stuck? Just assume to try with --cluster-alive-delete, if it fails try again without, and then move on?

Reply:
for openshift ci, 
```
destroy cluster --cluster-shutdown || true
destroy cluster
```

CI clusters will try to delete with shutdown, if it fails destroy all infra resources.
^^ this should succeed all the time , as requirement, for all types of openshift-CI(default)-clusters.

Reply:
Works for me.

Reply:
OK first I think we should require an explicit `ClusterDecomissioned` CRD that has something like a rationale attached, etc.

And then, `openshift-install destroy cluster` creates it, and waits for some part of the cluster to take ownership of it (probably machineAPI).  If machineAPI does, all the workers get forcibly killed, and then all control plane nodes except one, then finally the machineAPI running on the last machine kills itself.

If the machineAPI hasn't taken ownership of `ClusterDecomissioned`, then the installer can just do all of it.

Reply:
BTW whoever writes the "openshift-cluster-seppuku-operator" is going to have a LOT of fun writing it I'm sure :wink:

Reply:
It seems like the only way we maintain backwards compatibility is by requesting two stage destroy via a flag instead of by default based on these suggestions. I'll plan to add this to the enhancement.

Reply:
@cgwalters my problem with creating an object to signal deletion is the inability to know when everything which needed to respond has responded. How do I know that the machine api is going to respond to ClusterDecomissioned and i need to wait for it?

That's why finalizers on 'ClusterAlive' work. Because who needs to respond is known at the moment when the terminationDate is set.

Reply:
Perhaps this is a "why not both" situation.  As a cluster operator I can imagine it'd be extremely useful to have strong auditing around *why* a cluster is being destroyed.  If we had a rationale in the object I could imagine that e.g. the console and `oc status` would display it prominently.

And it's just generally useful to have an explicit object signaling the cluster will be destroyed - for example components like an autoscaling operator could check if it exists before spawning new machines, and skip doing so.

But the rationale for having finalizers for `ClusterAlive` does make sense, I just don't like the "implicit" nature of something blocking in a delete of that object as a signal for cluster deletion.